UnityAds/src/main/java/com/mopub/mobileads/UnityVideoAd.java

```diff
@@ -0,0 +1,153 @@
+import static com.mopub.mobileads.UnityAdsConstants.LOG_SHOW_ACTIVITY_NULL;
+import static com.mopub.mobileads.UnityAdsConstants.LOG_SHOW_AD_FAILURE;
+
+public abstract class UnityVideoAd extends UnityBaseAd {
+    private static final String ADAPTER_NAME = UnityVideoAd.class.getSimpleName();
+    private String mObjectId;
+
+    protected int impressionOrdinal;
+    protected int missedImpressionOrdinal;
+
+    protected abstract IUnityAdsShowListener getIUnityAdsShowListener();
+
+    @Override
+    protected void onInvalidate() { }
+
+    @Override
+    protected void load(@NonNull Context context, @NonNull AdData adData) {
+        super.load(context, adData);
+        if (failIfPlacementIdNull()) {
+            return;
+        }
+
+        MoPubLog.log(getAdNetworkId(), LOAD_ATTEMPTED, ADAPTER_NAME);
+        UnityAdsLoadOptions loadOptions = new UnityAdsLoadOptions();
+
+        String markup = adData.getExtras().get(DataKeys.ADM_KEY);
+        if (markup != null) {
+            loadOptions.setAdMarkup(markup);
+            mObjectId = UUID.randomUUID().toString();
+            loadOptions.setObjectId(mObjectId);
+        }
+        UnityAds.load(mPlacementId, loadOptions, mUnityLoadListener);
+    }
+
+    /**
+     * IUnityAdsLoadListener instance. Contains ad load success and fail logic.
+     */
+    private IUnityAdsLoadListener mUnityLoadListener = new IUnityAdsLoadListener() {
+        @Override
+        public void onUnityAdsAdLoaded(String placementId) {
+            mPlacementId = placementId;
+            MoPubLog.log(CUSTOM, ADAPTER_NAME, LOG_LOAD_SUCCESS + placementId);
+            MoPubLog.log(LOAD_SUCCESS, ADAPTER_NAME);
+
+            if (mLoadListener != null) {
+                mLoadListener.onAdLoaded();
+            }
+        }
+
+        @Override
+        public void onUnityAdsFailedToLoad(String placementId, UnityAds.UnityAdsLoadError error, String message) {
+            mPlacementId = placementId;
+            MoPubLog.log(CUSTOM, ADAPTER_NAME, LOG_LOAD_FAILED + placementId);
+            MoPubLog.log(LOAD_FAILED, ADAPTER_NAME, MoPubErrorCode.NETWORK_NO_FILL.getIntCode(), MoPubErrorCode.NETWORK_NO_FILL);
+
+            if (mLoadListener != null) {
+                mLoadListener.onAdLoadFailed(MoPubErrorCode.NETWORK_NO_FILL);
+            }
+        }
+    };
+
+    @Override
+    public void show() {
+        MoPubLog.log(SHOW_ATTEMPTED, ADAPTER_NAME);
+        MediationMetaData mediationMetaData = new MediationMetaData(mActivity);
+
+        if (mActivity == null || mPlacementId == null) {
+            if (mActivity == null) {
+                MoPubLog.log(CUSTOM, ADAPTER_NAME, LOG_SHOW_ACTIVITY_NULL);
+            }
+            if (mPlacementId == null) {
+                MoPubLog.log(CUSTOM, ADAPTER_NAME, LOG_PLACEMENT_ID_MISSING);
+            }
+            MoPubLog.log(SHOW_FAILED, ADAPTER_NAME, MoPubErrorCode.VIDEO_PLAYBACK_ERROR.getIntCode(), MoPubErrorCode.VIDEO_PLAYBACK_ERROR);
+
```

Comment:
Im looking at this, still burning my eyes, but I guess we wont be able to make it better.
The only question I have. is it ok that we are sending 2 logs, for example:
if Activity is null then we will send 
```java
MoPubLog.log(CUSTOM, ADAPTER_NAME, LOG_SHOW_ACTIVITY_NULL);
MoPubLog.log(SHOW_FAILED, ADAPTER_NAME, MoPubErrorCode.VIDEO_PLAYBACK_ERROR.getIntCode(), MoPubErrorCode.VIDEO_PLAYBACK_ERROR);
```

Reply:
Yeah I think it makes sense to report both, just like SDK metrics, they will measure the rate and could discover bugs this way